### PR TITLE
addition of ResourceValidation.ipynb and needed additions to imports.…

### PR DESCRIPTION
**Pull request by chrisroederucdenver**
…py. Also some polishing on existing imports.py

### CODE REVIEW

File: ResourceValidation.ipynb [hunk 1]
```diff
@@ -0,0 +1,202 @@
+{
```
Comment (chrisroederucdenver) [inline ResourceValidation.ipynb:1]: This is the start of some work to explore validation and validation errors. Very basic data-wise for now.

File: data/0_organization.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
+{
```
Comment (chrisroederucdenver) [inline data/0_organization.json:1]: This data directory is for the ResourceValidation work.

File: imports.py [hunk 3]
```diff
@@ -12,193 +10,283 @@
...
+        path = os.path.join(CACHE_DIR, filename)
+        if os.path.exists(path):
+            with open(path, 'r') as f:
+                data: Dict = json.load(f)
+                return data
+        return None
+
 
 
+
```
Comment (chrisroederucdenver) [inline imports.py:71]: All the lines are here because I removed tabs.

File: imports.py [hunk 4]
```diff
@@ -12,193 +10,283 @@
...
+class TimsClient:
+    """TIMS Client"""
+
+    def __init__(self, base_url: str, use_cache=False, verbose=False):
+        self.base_url = base_url
+        self.use_cache = use_cache
+        self.verbose=verbose
+        self.http_client = httpx.Client()
+        
+    def is_response_ok(self, response: Response) -> bool:
```
Comment (chrisroederucdenver) [inline imports.py:81]: I did what you made clear: flip the return value so we don't have the "not ...not_ok" double negative.

File: imports.py [hunk 5]
```diff
@@ -12,193 +10,283 @@
...
+            else:
+                print(f"error: {response.status_code}")
+                return False
+
+    def do_hapi_get_request(self, internal_path: str, in_params: Dict = None) -> Dict:
+        """GET request"""
+        url = self.base_url + internal_path
+        response_dict = None
+        if self.use_cache:
+            print(" * USING CACHE * ")
```
Comment (chrisroederucdenver) [inline imports.py:100]: I didn't want to see good results on accident, so I made it obvious when we're caching. The thinking is we should run from the server for real when saving the output.

File: imports.py [hunk 6]
```diff
@@ -12,193 +10,283 @@
...
+            if self.use_cache and self.verbose:
+                print("Cache empty for this request, fetching from server")
+            if self.verbose:
+                print("GETTING")
+            rq = httpx.Request("GET", url, params=in_params, headers=REQUEST_HEADERS)
+            if self.verbose:
+                print(rq.url)
+            response: Response = None
+            try:
+                if self.verbose:
```
Comment (chrisroederucdenver) [inline imports.py:114]: The server was being fussy and I got an exception here, so I added the catch. The exception type might be too wide...

File: imports.py [hunk 7]
```diff
@@ -12,193 +10,283 @@
...
+        url = self.base_url + internal_path
+        response_dict = None
+        if self.use_cache:
+            print(" * USING CACHE * ")
+            response_dict = CacheFhir.cache_read(url, in_params)
+        if response_dict:    
+            return response_dict
+        else:
+            if self.use_cache and self.verbose:
+                print("Cache empty for this request, fetching from server")
```
Comment (chrisroederucdenver) [inline imports.py:106]: covering the base when the cache doesn't have a file yet.

File: imports.py [hunk 8]
```diff
@@ -12,193 +10,283 @@
...
+            except Exception as x:
+                print(x)
+            if self.is_response_ok(response):        
+                response_dict: Dict = json.loads(response.text)
+                CacheFhir.cache_write(url, in_params, response_dict)
+                return response_dict
+            return None
+
+    def do_hapi_post_request(self, internal_path, form_data, in_params) -> Dict:
+        rq = httpx.Request("POST", self.base_url + internal_path,
```
Comment (chrisroederucdenver) [inline imports.py:126]: needed for ResourceValidation

File: imports.py [hunk 9]
```diff
@@ -12,193 +10,283 @@
...
+                    print("SENDING")
+                response: Response = self.http_client.send(rq)
+            except Exception as x:
+                print(x)
+            if self.is_response_ok(response):        
+                response_dict: Dict = json.loads(response.text)
+                CacheFhir.cache_write(url, in_params, response_dict)
+                return response_dict
+            return None
+
```
Comment (chrisroederucdenver) [inline imports.py:124]: worth noticing this returns a None when things didn't work. Checks below.

File: imports.py [hunk 10]
```diff
@@ -12,193 +10,283 @@
...
+        ## if self.is_response_ok(response):        
+        response_dict: Dict = json.loads(response.text)
+        return response_dict
+   
+
+    def summarize_code_systems(self, verbose=False):
+        """Summarize code systems"""
+        response = self.do_hapi_get_request("CodeSystem?_summary=true", {})
+        if response:        
+            sys_names = [x['resource']['name'] for x in response['entry']]
```
Comment (chrisroederucdenver) [inline imports.py:142]: here's the check for None, so we don't blow exceptions if the request was bad.

File: imports.py [hunk 11]
```diff
@@ -12,193 +10,283 @@
...
 DEBUG = False
 if __name__ == '__main__':
-	if DEBUG:
-		client = TimsClient(base_url=BASE_URL, use_cache=True)
-		# examples:
-		client.summarize_code_system_by_name('LOINC')
-		client.lookup_code(system='http://loinc.org', code='LA6668-3')
+    if DEBUG:
+        BASE_URL = "http://20.119.216.32:8000/r4/"  # TIMS server for 
+        client = TimsClient(base_url=BASE_URL, use_cache=True)
```
Comment (chrisroederucdenver) [inline imports.py:304]: I though the BASE_URL here was a little confusing with the one passed in, so I scooted it down here to be closer to the scope where it's used.

File: imports.py [hunk 12]
```diff
@@ -12,193 +10,283 @@
...
-			print(label)
-		else:
-			print('Invalid code')
-
-	def lookup_code(self, system: URI, code: str):
-		"""Lookup code within code system"""
-		response: Dict = self.do_hapi_request(f'CodeSystem/$lookup?system={system}&code={code}')
-		decode_lookup_response(response)
+class CacheFhir:
+    """ missing an invalidation mechanism, use with caution """
```
Comment (chrisroederucdenver) [inline imports.py:22]: Since you opened the OO here, I thought this made sense. If only to group them.

File: imports.py [hunk 13]
```diff
@@ -12,193 +10,283 @@
...
+        self.http_client = httpx.Client()
+        
+    def is_response_ok(self, response: Response) -> bool:
+        """Decode status"""
+        if response is None:
+            print("NO RESPONSE")
+            return False
+        else:
+            if  200 <= response.status_code < 300:
+                if self.verbose:
```
Comment (chrisroederucdenver) [inline imports.py:88]: I wasn't using the break-out for different types of errors, so simplified this.

File: imports.py (hunk 12 above)
Comment (chrisroederucdenver) [reply imports.py:22]: Also, I see why you set this up. The JHU server, or the net between here and there, isn't nearly as fast as a local server. Local just takes a few milli seconds. JHU could be 10 seconds. I hope the memory helps it out.

**Review by chrisroederucdenver — COMMENTED** [review]
added code comments to help review

File: imports.py (hunk 3 above)
Comment (joeflack4) [reply imports.py:71]: ah i see

File: imports.py (hunk 4 above)
Comment (joeflack4) [reply imports.py:81]: thanks!

File: imports.py (hunk 5 above)
Comment (joeflack4) [reply imports.py:100]: Fine with this for now. My thought is just to have as little on the screen as possible to demonstrate the functionality.  Of course I don't want to rely on the cache simply as a workaround for long or buggy server responses; that would be just wrong. Mostly I have it here for debugging purposes anyway, so I suppose that for future 'release' PDFs (or when we save the notebook itself), it'll be w/ caching off.

File: imports.py (hunk 6 above)
Comment (joeflack4) [reply imports.py:114]: yeah, too wide for now, but that's ok.

i was surprised that so many of my ad hoc attempts failed when trying to demonstrate the notebook live today. im guessing that in a lot of cases it wasn't that I got an error back, but that the JSON coming back was a different shape than expected. not sure

in any case, I don't aim to make the code behind the notebook into a super robust client. I'm pretty happy with HAPI's swagger UI for that

File: imports.py (hunk 8 above)
Comment (joeflack4) [reply imports.py:126]: ah ok, I'm not too familiar w/ that but I'm glad you're looking into it!

File: ResourceValidation.ipynb (hunk 1 above)
Comment (joeflack4) [reply ResourceValidation.ipynb:1]: nice. send me an update when it's ready for me to look at visually!

**Comment by joeflack4** [comment]
@chrisroederucdenver Thanks for all the explanations! very detailed

**Pull request merged**